activerecord/lib/active_record/relation/calculations.rb

```diff
@@ -348,8 +348,8 @@ def column_alias_for(keys)
     end
 
     def type_for(field)
-      field_name = field.respond_to?(:name) ? field.name.to_s : field.to_s.split('.').last
-      @klass.type_for_attribute(field_name)
+      field = field.name if field.respond_to?(:name)
+      @klass.columns_hash[field.to_s]
```

Comment:
The `columns_hash` no longer contains appropriate type information. `type_for_attribute` is the correct method to use.